Add asset info to library guide (#34217)
angular/angular: aio/content/guide/creating-libraries.md

If you've never published a package in npm before, you must create a user account. Read more in [Publishing npm Packages](https://docs.npmjs.com/getting-started/publishing-npm-packages).

{@a lib-assets}

### Including assets with libraries

> The [ng-packagr](https://github.com/ng-packagr/ng-packagr/blob/master/README.md) builder can automatically copy assets into your library package as part of the build process.

Reviewer: - ng-packagr is not a builder, it’s a tool
- @angular-devkit/ng-packagr is a builder

> Learn more about how to [embed assets in CSS](https://github.com/ng-packagr/ng-packagr/blob/master/docs/embed-assets-css.md).

Reviewer: This doc is related to a bit of a different topic. It’s how to embed an asset in a style sheet.

For the “feature” that is mentioned here, there are no docs yet.

Author: @alan-agius4 Have you had a chance to look at this yet?

Reviewer: No sorry didn’t have time yet. Will try to allocate some time tomorrow.

Reviewer: FYI: Assets and embedded assets within css are 2 different features, which serve different purposes.

The former is used for assets which are NOT referenced in Angular components templates or stylesheets. Typically such assets are `.scss` files that contain mixins which can be used for theming. While the latter is used to embed assets which are referenced in Angular Component templates and styles.

Author: I'm not familiar with this area - can you give me more direction about what specifically needs to  change, or needs to be said?

Reviewer: I guess, it's better to amend this once we have the "assets" documentation in ng-packagr, so that we can link it. I'll try to write up something in the coming days.

If you've never published a package in npm before, you must create a user account. Read more in [Publishing npm Packages](https://docs.npmjs.com/getting-started/publishing-npm-packages).

{@a lib-assets}

> ### Managing library assets with ng-packagr

Reviewer 2: this section has nothing to do with "publishing your library" so it shouldn't be the subsection of this section. I suggest we move it to the very bottom of this md file as a subsection of "Building and rebuilding your library"

Author: Thanks @IgorMinar 
I will remove the "TBD" when the new page is added to `ng-packagr`

Reviewer 3: I think that this section on publishing apps with additional assets is related to "publishing your library". This is the step of library development where library authors run into the issue of needing to copy assets during build time into their production bundle for publishing on NPM. This often happens when an already published library wants to add theming mixins to their published library.

I think adding this to a separate "building your library" section could work, but currently the "getting started" section covers `ng build my-lib` and the "publishing your library" section covers `ng build my-lib --prod`.

The current hierarchy doesn't seem to make sense
![Screen Shot 2020-02-03 at 14 02 13](https://user-images.githubusercontent.com/3506071/73682463-6c083900-468e-11ea-9827-e183565483c7.png)

It doesn't seem like "Managing library assets with ng-packagr" should be nested under the "Using your own library in apps" section.

Starting with version 9.x of the [ng-packagr](https://github.com/ng-packagr/ng-packagr/blob/master/README.md) tool, you can use the tool to automatically copy assets into your library package as part of the build process.

> * Learn how to [copy assets into your library as part of the build](https://github.com/ng-packagr/ng-packagr/blob/master/docs/copy-assets-css.md). *TBD*

Reviewer 2: this link doesn't exit, and "TBD" should not be part of our content. Should we remove this or update the link and remove TBD?

Author: That's why this is "blocked" -- waiting for Alan to create this page.

This mapping ensures that your library always loads the local copies of the modules it needs.

{@a lib-assets}

> ## Including assets with libraries

Reviewer: From devrel perspective how do we usually communicate something like this?

That being, while this is documented it’s not something that we support.

//cc @mgechev

Reviewer 3: In what sense is this "not supported"? You just want to call out that `scss-bundle` is not owned and maintained by the Angular team?

Reviewer 4: We can add a warning box stating that this is not supported by the team, but an approach provided by a third-party library.

Reviewer: Something like that, my point is that using a package might cause undefined behaviour, example unless changed using this package will opt you out of using sass and will force you to use the legacy node-sass since it has direct dependency on node-sass. This will have a effect on all workspace projects which might break currently working project. 

Ps: will need to confirm if they still use node sass.

Also, afaik previously we reframing from adding certain documents in aio for “unsupported” reasons, such as HMR, bootstrap etc...

Reviewer: Maybe we should revisit and add documents such as how to configure HMR etc...

> As a library author, you can process style files in any of the supported formats (CSS, SCSS, or SASS) and generate CSS files to be included with your library.

Reviewer 3: We may not want to list the formats here since I believe there are other supported formats than what we list here.

Reviewer 3: This isn't about generating CSS, it's about bundling SCSS files with your library. These can then be consumed by applications which will handle the generation of CSS from the combination of their SCSS and the library's bundled SCSS.

Reviewer 5: I agree that it's better to not enumerate the formats here and instead link to the canonical docs for support style formats.

There are various ways to do this.

> Angular Material, for example, uses a utility called [scss-bundle](https://www.npmjs.com/package/scss-bundle) to generate multiple CSS files.

Reviewer 3: ```suggestion
Angular Material, for example, uses a utility called [scss-bundle](https://www.npmjs.com/package/scss-bundle) to include multiple SCSS theme bundles with their published library.
```

Reviewer 5: This also isn't exactly right; we use `scss-bundle` to combine _multiple_ Sass partial files into _one_ Sass partial file. Our actual theme css files are just compiled directly with the Sass compiler (via Bazel).

> Angular Material, for example, uses a utility called [scss-bundle](https://www.npmjs.com/package/scss-bundle) to generate multiple CSS files.

Reviewer 5: I'm iffy about whether we want to call attention to this. We only use `scss-bundle` at this point because it improves the performance of downstream Sass compilations (since the compiler doesn't have to process multiple files). However, the Sass team is working on fixing performance issues around file processing, so this may become unnecessary within the next six months. 

For most libraries, they'd probably be fine just including their Sass sources in the package output directly.

Reviewer: I think i agree with this. I think what should be done is that we should add something similar to `srcs` attribute of the `npm_package` Bazel rule to `ng-packagr` the package to allow basic copying of files. 

This should be trivial to implement.

//cc @Splaktar what do you think?

Reviewer 3: Oh I agree that it would be much preferred to add support for this to ng-packagr rather than documenting a workaround! However, the issue has been open for a couple years now with ng-packagr and doesn't appear to be getting any attention (other than people asking for it).

Reviewer: Now it has 😜 can you send me the issue number over slack? 

If it’s kept simple I’ll should be able to do something in the next couple of days.

Reviewer: https://github.com/ng-packagr/ng-packagr/pull/1465

> To use this tool, load the npm package and create a configuration file named `scss-bundle.config.json`.

Reviewer: I don’t think this description is correct. Scss-bundle doesn’t generate an CSS. But rather bundles SCSS and Sass into a single file.

> The configuration tells the tool where to get the bundle and where to put the generated CSS, and lets you provide options for how to generate the styles.

Reviewer 3: ```suggestion
Learn more about [`scss-bundle` configuration](https://github.com/reactway/scss-bundle#configuration).
```

> The options in this file can be overridden by similar options in the CLI configuration file, `angular.json`. For details, see the [package documentation](https://www.npmjs.com/package/scss-bundle#configuration).

Reviewer: I am not following why this line is needed and comparing it with angular.json

They are 2 unrelated tools.

Reviewer 3: Yeah, I think that this line (169) should be removed.

> The options in this file can be overridden by similar options in the CLI configuration file, `angular.json`. For details, see the [package documentation](https://www.npmjs.com/package/scss-bundle#configuration).

Reviewer 3: ```suggestion
```

When you have created the configuration file, add a script to invoke the tool.

<code-example language="html" header="??name of file??">

> "build-css" : "scss-bundle -p $lt;project-path-to-scss-bundle.config.json>"

Reviewer 3: ```suggestion
"build:scss-bundle" : "scss-bundle -p $lt;project-path-to-scss-bundle.config.json>"
```

</code-example>

> A [custom CLI builder](https://github.com/linnenschmidt/build-ng-packagr) is also available that will automatically copy the assets folder created by the bundler into your library package, as part of the build process.

Reviewer 3: I'm not sure that we should include this here as it isn't directly related to `scss-bundle` and isn't as well supported/established.

Reviewer: I agree, and using a custom builder has its cavets.

The library author will need to create migrations script for each majoy angular version.

This is because certain cli migrations will only be executed if the builders are from @angular-devkit/build-angular. The main reason for this is because if we don’t do this we might break users since custom builders might have all sort of options which we don’t know of.

TypeScript path mappings should *not* point to the library source `.ts` files.

</div>

{@a lib-assets}

### Managing library assets with ng-packagr

> Starting with version 9.x of the [ng-packagr](https://github.com/ng-packagr/ng-packagr/blob/master/README.md) tool, you can use the tool to automatically copy assets into your library package as part of the build process.

Reviewer 3: ```suggestion
Starting with version 9.x of the [ng-packagr](https://github.com/ng-packagr/ng-packagr/blob/master/README.md) tool, you can configure the tool to automatically copy assets into your library package as part of the build process. You can use this feature when your library needs to publish optional theming files, Sass Mixins, or documentation (like a change log).
```

Reviewer 6: Hey @Splaktar, out of curiosity, why just sass mixins? Does it not include other sass features?

Reviewer 3: Sass mixins are most likely to be provided, but not used by the library. They may be provided to allow the consumer to optionally override a theme or other behavior of the app where just overriding a single variable is not enough. If the user doesn't need to customize this behavior, they don't have to import these mixins and can save on bundle size.

Other Sass features are most likely to be used by the library directly and thus bundled into the build by the CLI.

Reviewer 3: Note that a library likely has lots of other Sass mixins that it uses. So there may be a split of required and optional mixin bundles.

Reviewer 6: Ah, thank you for the clarification, @Splaktar.

TypeScript path mappings should *not* point to the library source `.ts` files.

</div>

{@a lib-assets}

### Managing library assets with ng-packagr

Starting with version 9.x of the [ng-packagr](https://github.com/ng-packagr/ng-packagr/blob/master/README.md) tool, you can configure the tool to automatically copy assets into your library package as part of the build process.

> You can use this feature when your library needs to publish optional theming files, Sass mixins, or documentation (like a change log).

Reviewer: Not: change log -> changelog
